### PR TITLE
History tests

### PR DESCRIPTION
**Pull request by edynox**
Added some tests for dnf `history undo` and `redo`

### CODE REVIEW

File: dnf-docker-test/features/steps/test_behave.py [hunk 1]
```diff
@@ -263,3 +263,21 @@ def then_the_section(context, section, pkgs):
                     return
     except StopIteration:
         raise AssertionError('section {} has not been found'.format(section))
+
```
Comment (ignatenkobrain) [inline dnf-docker-test/features/steps/test_behave.py]: please, nothing in test_behave. it's deprecated

Comment (edynox) [inline dnf-docker-test/features/steps/test_behave.py]: Sorry I didn't know that. Will fix it...

Comment (edynox) [inline dnf-docker-test/features/steps/test_behave.py]: should be fine for now... I will probably add some more this week.

File: dnf-docker-test/features/history-1.feature [hunk 2]
```diff
@@ -0,0 +1,72 @@
+Feature: DNF/Behave test Transaction history - base
+
+    Scenario: Simple transaction
+        Given I use the repository "test-1"
```
Comment (ignatenkobrain) [inline dnf-docker-test/features/history-1.feature]: This step is deprecated

File: dnf-docker-test/features/history-1.feature [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        When I execute "dnf" command "install -y TestA" with "success"
```
Comment (ignatenkobrain) [inline dnf-docker-test/features/history-1.feature]: This step is deprecated

Comment (edynox) [inline dnf-docker-test/features/history-1.feature]: Are there any alternatives for these two? Or should I write a new ones?

Comment (ignatenkobrain) [inline dnf-docker-test/features/history-1.feature]: there is, check `obsoletes-reason` for example

**Comment by edynox** [comment]
Should be fine now

File: dnf-docker-test/features/history-1.feature [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+          And I successfully run "dnf history redo last-1 -y"
+        Then rpmdb changes are
+          | State        | Packages      |
+          | removed      | TestA, TestB  |
+        And history should contain "history redo last-1 -y" with action "Erase" and "2" packages
+
+    Scenario: Update packages
+        When I save rpmdb
+          And I successfully run "dnf install -y TestA TestB TestD"
+        When I save rpmdb
```
Comment (ignatenkobrain) [inline dnf-docker-test/features/history-1.feature]: you should first check if all those packages have been installed correctly

File: dnf-docker-test/features/history-1.feature [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    Scenario: Undo last transaction
+        When I save rpmdb
+          And I successfully run "dnf history undo last -y"
+        Then rpmdb changes are
+          | State        | Packages      |
+          | removed      | TestA, TestB  |
+        And history should contain "history undo last -y" with action "Erase" and "2" packages
+
+    Scenario: Undo transaction last 2
```
Comment (ignatenkobrain) [inline dnf-docker-test/features/history-1.feature]: last 2?

File: dnf-docker-test/features/history-1.feature (hunk 4 above)
Comment (edynox) [inline dnf-docker-test/features/history-1.feature]: fixed

File: dnf-docker-test/features/history-1.feature (hunk 5 above)
Comment (edynox) [inline dnf-docker-test/features/history-1.feature]: is ` Undo last transaction 2` better? :smile:

**Comment by edynox** [comment]
History tests should be complete now.

**Comment by edynox** [comment]
fixed indentation

**Comment by j-mracek** [comment]
@edynox I don't know why but test history-2 is failing. Please can you investigate it or provide an explanation?

**Comment by edynox** [comment]
@j-mracek works fine with SWDB :smile: 
Can you provide me a output?
If it is like `AttributeError: 'list' object has no attribute 'add'`
then https://github.com/rpm-software-management/dnf/pull/756 should fix it

**Comment by j-mracek** [comment]
Here it is:
+ behave-2 -i history-2-dnf-3 -D dnf_cmd=dnf-3 --junit --junit-directory /junit/ /behave/
Feature: DNF/Behave test Transaction history [info, list, userinstalled] # behave/history-2-dnf-3.feature:1

  @setup
  Scenario: Feature Setup                  # behave/history-2-dnf-3.feature:4
    Given repository "test" with packages  # behave/steps/repo_steps.py:228
      | Package | Tag      | Value |
      | TestA   | Requires | TestB |
      | TestB   |          |       |
      | TestC   |          |       |
      | TestD   | Requires | TestE |
      | TestE   | Requires | TestC |
    And repository "updates" with packages # behave/steps/repo_steps.py:228
      | Package | Tag     | Value |
      | TestA   | Version | 2     |
      | TestB   | Version | 2     |
      | TestC   | Version | 2     |
      | TestD   | Version | 2     |
    When I save rpmdb                      # behave/steps/rpm_steps.py:15
    And I enable repository "test"         # behave/steps/repo_steps.py:417

  Scenario: List userinstalled packages                 # behave/history-2-dnf-3.feature:21
    When I save rpmdb                                   # behave/steps/rpm_steps.py:15
    And I successfully run "dnf install -y TestA TestE" # behave/steps/command_steps.py:33
    Then history userinstalled should                   # behave/steps/command_steps.py:145
      | Action    | Packages     |
      | Match     | TestA, TestE |
      | Not match | TestB, TestC |

  Scenario: History list last and range                                                          # behave/history-2-dnf-3.feature:29
    When I save rpmdb                                                                            # behave/steps/rpm_steps.py:15
    And I successfully run "dnf install -y TestD"                                                # behave/steps/command_steps.py:33
    And I successfully run "dnf remove -y TestE"                                                 # behave/steps/command_steps.py:33
    Then history "last-1..last" should match "install -y TestD" with "Install" and "1" package   # behave/steps/command_steps.py:124
    And history "list last-1..last" should match "remove -y TestE" with "Erase" and "3" packages # behave/steps/command_steps.py:124
    And history "last" should match "remove -y TestE" with "Erase" and "3" packages              # behave/steps/command_steps.py:124

  Scenario: History list package                                                          # behave/history-2-dnf-3.feature:37
    When I save rpmdb                                                                     # behave/steps/rpm_steps.py:15
    And I successfully run "dnf install -y TestC"                                         # behave/steps/command_steps.py:33
    Then history "TestC" should match "install -y TestC" with "Install" and "1" package   # behave/steps/command_steps.py:124
    And history "list TestC" should match "remove -y TestE" with "Erase" and "3" packages # behave/steps/command_steps.py:124

  Scenario: History info                          # behave/history-2-dnf-3.feature:43
    When I save rpmdb                             # behave/steps/rpm_steps.py:15
    Then history info should match                # behave/steps/command_steps.py:166
      | Key          | Value            |
      | Command Line | install -y TestC |
      | Return-Code  | Success          |
      | Install      | TestC            |
      Assertion Failed: Begin rpmdb version not found
      Captured stdout:
      shell.command: ['dnf-3', 'history', 'info']
      shell.command.stdout:
      Last metadata expiration check: 0:00:04 ago on Fri May 12 06:55:54 2017 UTC.
      
      shell.command.stderr:
      AttributeError: 'list' object has no attribute 'add'

    When I successfully run "dnf remove -y TestA" # None
    Then history info should match                # None
      | Key          | Value           |
      | Command Line | remove -y TestA |
      | Return-Code  | Success         |
      | Erase        | TestA, TestB    |

  Scenario: History info in range                 # behave/history-2-dnf-3.feature:57
    When I save rpmdb                             # behave/steps/rpm_steps.py:15
    And I successfully run "dnf install -y TestA" # behave/steps/command_steps.py:33
    Then history info should match                # behave/steps/command_steps.py:166
      | Key          | Value            |
      | Command Line | install -y TestA |
      | Return-Code  | Success          |
      | Install      | TestA, TestB     |
      Assertion Failed: Begin rpmdb version not found
      Captured stdout:
      shell.command: ['dnf-3', 'install', '-y', 'TestA']
      shell.command.stdout:
      Last metadata expiration check: 0:00:04 ago on Fri May 12 06:55:54 2017 UTC.
      Dependencies resolved.
      Nothing to do.
      Complete!
      
      shell.command.stderr:
      Package TestA-1-1.noarch is already installed, skipping.
      
      shell.command: ['dnf-3', 'history', 'info']
      shell.command.stdout:
      Last metadata expiration check: 0:00:04 ago on Fri May 12 06:55:54 2017 UTC.
      
      shell.command.stderr:
      AttributeError: 'list' object has no attribute 'add'

    When I save rpmdb                             # None
    And I enable repository "updates"             # None
    And I successfully run "dnf update -y"        # None
    Then history info should match                # None
      | Key          | Value        |
      | Command Line | update -y    |
      | Return-Code  | Success      |
      | Upgrade      | TestA, TestB |
      | Upgraded     | TestA, TestB |
    And history info "last-2..last" should match  # None
      | Key         | Value        |
      | Return-Code | Success      |
      | Upgrade     | TestA, TestB |
      | Upgraded    | TestA, TestB |
      | Install     | TestA, TestB |
      | Erase       | TestA, TestB |
    And history info "TestA" should match         # None
      | Key         | Value                      |
      | Return-Code | Success                    |
      | Upgrade     | TestA, TestB               |
      | Upgraded    | TestA, TestB               |
      | Install     | TestA, TestB, TestC, TestE |
      | Erase       | TestA, TestB               |


Failing scenarios:
  behave/history-2-dnf-3.feature:43  History info
  behave/history-2-dnf-3.feature:57  History info in range

0 features passed, 1 failed, 0 skipped
4 scenarios passed, 2 failed, 0 skipped
20 steps passed, 2 failed, 8 skipped, 0 undefined

**Comment by j-mracek** [comment]
I test it with upstream dnf, but as I good understand it requires your database swdb. Is it like that?

**Comment by edynox** [comment]
Then history info is broken in upstream
```
      shell.command.stderr:
      AttributeError: 'list' object has no attribute 'add'
```
I have the same output in rawhide docker image
```
[xcubae00@ares ~]$ docker run -it --rm fedora:rawhide bash
[root@40ff792a3361 /]# dnf history info
Fedora - Rawhide - Developmental packages for t  11 MB/s |  54 MB     00:04    
Last metadata expiration check: 0:00:20 ago on Fri May 12 07:36:36 2017 UTC.
AttributeError: 'list' object has no attribute 'add'
[root@40ff792a3361 /]# 
```

Like I said, https://github.com/rpm-software-management/dnf/pull/756 should fix that issue
History tests aren't dependent on SWDB.

**Comment by j-mracek** [comment]
@edynox Still one step failing even with your patch:
```
 Then history info should match                # behave/steps/command_steps.py:166
      | Key          | Value        |
      | Command Line | update -y    |
      | Return-Code  | Success      |
      | Upgrade      | TestA, TestB |
      | Upgraded     | TestA, TestB |
    And history info "last-2..last" should match  # behave/steps/command_steps.py:166
      | Key         | Value        |
      | Return-Code | Success      |
      | Upgrade     | TestA, TestB |
      | Upgraded    | TestA, TestB |
      | Install     | TestA, TestB |
      | Erase       | TestA, TestB |
      Assertion Failed: "TestA" not matched as "Install"
      Captured stdout:
      shell.command: ['dnf-2', 'install', '-y', 'TestA']
      shell.command.stdout:
      Last metadata expiration check: 0:00:04 ago on Wed May 17 11:40:10 2017 UTC.
      Dependencies resolved.
      ================================================================================
       Package           Arch               Version            Repository        Size
      ================================================================================
      Installing:
       TestA             noarch             1-1                test             5.5 k
      Installing dependencies:
       TestB             noarch             1-1                test             5.4 k
      
      Transaction Summary
      ================================================================================
      Install  2 Packages
      
      Total size: 11 k
      Installed size: 0  
      Running transaction check
      Transaction check succeeded.
      Running transaction test
      Transaction test succeeded.
      Running transaction
        Preparing        :                                                        1/1 
        Installing       : TestB-1-1.noarch                                       1/2 
        Installing       : TestA-1-1.noarch                                       2/2 
        Verifying        : TestA-1-1.noarch                                       1/2 
        Verifying        : TestB-1-1.noarch                                       2/2 
      
      Installed:
        TestA.noarch 1-1                       TestB.noarch 1-1                      
      
      Complete!
      
      shell.command.stderr:
      
      shell.command: ['dnf-2', 'history', 'info']
      shell.command.stdout:
      Last metadata expiration check: 0:00:05 ago on Wed May 17 11:40:10 2017 UTC.
      Transaction ID : 10
      Begin time     : Wed May 17 11:40:15 2017
      Begin rpmdb    : 252:66f6c164c5f498979cdce259868e19d90c855b66
      End time       :                           (0 seconds)
      End rpmdb      : 254:4543a7d0514c0c9f6e516d0b2aff50ed8619d6ab
      User           : System <unset>
      Return-Code    : Success
      Command Line   : install -y TestA
      Transaction performed with:
          Installed     dnf-2.4.2-0.4g139e6e8.fc25.noarch @rpmsoftwaremanagement-dnf-nightly
          Installed     rpm-4.13.0.1-1.fc25.x86_64        @koji-override-0
      Packages Altered:
          Install TestA-1-1.noarch @test
          Install TestB-1-1.noarch @test
      
      shell.command.stderr:
      
      shell.command: ['dnf-2', 'update', '-y']
      shell.command.stdout:
      Last metadata expiration check: 0:00:00 ago on Wed May 17 11:40:16 2017 UTC.
      Dependencies resolved.
      ================================================================================
       Package           Arch               Version         Repository           Size
      ================================================================================
      Upgrading:
       TestA             noarch             2-1             updates             5.4 k
       TestB             noarch             2-1             updates             5.4 k
       TestC             noarch             2-1             updates             5.4 k
      
      Transaction Summary
      ================================================================================
      Upgrade  3 Packages
      
      Total size: 16 k
      Running transaction check
      Transaction check succeeded.
      Running transaction test
      Transaction test succeeded.
      Running transaction
        Preparing        :                                                        1/1 
        Upgrading        : TestC-2-1.noarch                                       1/6 
        Upgrading        : TestB-2-1.noarch                                       2/6 
        Upgrading        : TestA-2-1.noarch                                       3/6 
        Cleanup          : TestA-1-1.noarch                                       4/6 
        Cleanup          : TestB-1-1.noarch                                       5/6 
        Cleanup          : TestC-1-1.noarch                                       6/6 
        Verifying        : TestA-2-1.noarch                                       1/6 
        Verifying        : TestB-2-1.noarch                                       2/6 
        Verifying        : TestC-2-1.noarch                                       3/6 
        Verifying        : TestA-1-1.noarch                                       4/6 
        Verifying        : TestB-1-1.noarch                                       5/6 
        Verifying        : TestC-1-1.noarch                                       6/6 
      
      Upgraded:
        TestA.noarch 2-1          TestB.noarch 2-1          TestC.noarch 2-1         
      
      Complete!
      
      shell.command.stderr:
      
      shell.command: ['dnf-2', 'history', 'info']
      shell.command.stdout:
      Last metadata expiration check: 0:00:01 ago on Wed May 17 11:40:16 2017 UTC.
      Transaction ID : 11
      Begin time     : Wed May 17 11:40:16 2017
      Begin rpmdb    : 254:4543a7d0514c0c9f6e516d0b2aff50ed8619d6ab
      End time       :            11:40:17 2017 (1 seconds)
      End rpmdb      : 254:4780c0c732cdc98c66f4143aac080f52d87a5d8c
      User           : System <unset>
      Return-Code    : Success
      Command Line   : update -y
      Transaction performed with:
          Installed     dnf-2.4.2-0.4g139e6e8.fc25.noarch @rpmsoftwaremanagement-dnf-nightly
          Installed     rpm-4.13.0.1-1.fc25.x86_64        @koji-override-0
      Packages Altered:
          Upgraded TestA-1-1.noarch @test
          Upgrade        2-1.noarch @updates
          Upgraded TestB-1-1.noarch @test
          Upgrade        2-1.noarch @updates
          Upgraded TestC-1-1.noarch @test
          Upgrade        2-1.noarch @updates
      
      shell.command.stderr:
      
      shell.command: ['dnf-2', 'history', 'info', 'last-2..last']
      shell.command.stdout:
      Last metadata expiration check: 0:00:01 ago on Wed May 17 11:40:16 2017 UTC.
      Transaction ID : 9..11
      Begin time     : Wed May 17 11:40:14 2017
      Begin rpmdb    : 254:4543a7d0514c0c9f6e516d0b2aff50ed8619d6ab
      End time       :            11:40:17 2017 (3 seconds)
      End rpmdb      : 254:4780c0c732cdc98c66f4143aac080f52d87a5d8c
      User           : System <unset>
      Return-Code    : Success
      Command Line   : remove -y TestA
      Command Line   : install -y TestA
      Command Line   : update -y
      Transaction performed with:
          Installed     dnf-2.4.2-0.4g139e6e8.fc25.noarch @rpmsoftwaremanagement-dnf-nightly
          Installed     rpm-4.13.0.1-1.fc25.x86_64        @koji-override-0
      Packages Altered:
          Upgraded TestA-1-1.noarch @test
          Upgrade        2-1.noarch @updates
          Upgraded TestB-1-1.noarch @test
          Upgrade        2-1.noarch @updates
          Upgraded TestC-1-1.noarch @test
          Upgrade        2-1.noarch @updates
      
      shell.command.stderr:


```

**Comment by j-mracek** [comment]
@edynox Please can you fix that it will work with upstream version? Then I can merge it. Thanks a lot. Please if there is any problem, just let me know.

**Comment by edynox** [comment]
@j-mracek 
I have transaction merging problem in SWDB.
Now I'm rewriting YumMergedHistoryTransaction into libdnf. 
DnfSwdbTrans will support merge() method and we can drop both YumMergedHistoryTransaction and YumHistoryTransaction from yum/history.py.

Then upstream and SWDB output should be the same and I will rewrite this test so transaction merging will be tested properly (and passed by both upstream an SWDB)

It is quite lot of work with proper transaction merging so I will need a few days...

**Comment by edynox** [comment]
https://github.com/rpm-software-management/libdnf/pull/199/commits/509f859ac3b982e54b5eab28832826aec53f47e3 introduces transaction merging in SWDB, I will write some tests on it here. YumMergedHistoryTransaction and YumHistoryTransaction were dropped from dnf.

**Comment by edynox** [comment]
@j-mracek Merged transactions should be tested properly now. Its working on upstream for me.

**Comment by edynox** [comment]
@j-mracek Can you finish the review, please?

**Comment by j-mracek** [comment]
@edynox I am sorry but I got:
```
Failed test(s): history-2-dnf-2 history-2-dnf-3
```
Second time that I repeated it, it passed (same container), and first test output is not reachable.

The reason has be found first.

**Comment by j-mracek** [comment]
```
sudo docker run  --rm dnf-bot/dnf-testing:latest launch-test  history-2 dnf-2
+ RESERVE=0
+ '[' history-2 == -r -o history-2 == -R ']'
+ httpd -k start
+ vsftpd
+ rngd -r /dev/urandom -o /dev/random
unable to adjust write_wakeup_threshold: Read-only file system
+ new_name=history-2-dnf-2
+ EXIT_STATUS=0
+ cp /tests/history-2.feature /behave/history-2-dnf-2.feature
+ '[' 0 = 1 ']'
+ TEST_EXIT=0
+ behave-2 --tags '~@backup' --tags '~@restore' -i history-2-dnf-2 -D dnf_cmd=dnf-2 --junit --junit-directory /junit/ /behave/
Feature: DNF/Behave test Transaction history [info, list, userinstalled] # behave/history-2-dnf-2.feature:1

  @setup
  Scenario: Feature Setup                  # behave/history-2-dnf-2.feature:4
    Given repository "test" with packages  # behave/steps/repo_steps.py:254
      | Package | Tag      | Value |
      | TestA   | Requires | TestB |
      | TestB   |          |       |
      | TestC   |          |       |
      | TestD   | Requires | TestE |
      | TestE   | Requires | TestC |
    And repository "updates" with packages # behave/steps/repo_steps.py:254
      | Package | Tag     | Value |
      | TestA   | Version | 2     |
      | TestB   | Version | 2     |
      | TestC   | Version | 2     |
      | TestD   | Version | 2     |
    When I save rpmdb                      # behave/steps/rpm_steps.py:16
    And I enable repository "test"         # behave/steps/repo_steps.py:488

  Scenario: List userinstalled packages                 # behave/history-2-dnf-2.feature:21
    When I save rpmdb                                   # behave/steps/rpm_steps.py:16
    And I successfully run "dnf install -y TestA TestE" # behave/steps/command_steps.py:33
    Then history userinstalled should                   # behave/steps/command_steps.py:174
      | Action    | Packages     |
      | Match     | TestA, TestE |
      | Not match | TestB, TestC |

  Scenario: History list last and range                                                          # behave/history-2-dnf-2.feature:29
    When I save rpmdb                                                                            # behave/steps/rpm_steps.py:16
    And I successfully run "dnf install -y TestD"                                                # behave/steps/command_steps.py:33
    And I successfully run "dnf remove -y TestE"                                                 # behave/steps/command_steps.py:33
    Then history "last-1..last" should match "install -y TestD" with "Install" and "1" package   # behave/steps/command_steps.py:153
    And history "list last-1..last" should match "remove -y TestE" with "Erase" and "3" packages # behave/steps/command_steps.py:153
    And history "last" should match "remove -y TestE" with "Erase" and "3" packages              # behave/steps/command_steps.py:153

  Scenario: History list package                                                          # behave/history-2-dnf-2.feature:37
    When I save rpmdb                                                                     # behave/steps/rpm_steps.py:16
    And I successfully run "dnf install -y TestC"                                         # behave/steps/command_steps.py:33
    Then history "TestC" should match "install -y TestC" with "Install" and "1" package   # behave/steps/command_steps.py:153
    And history "list TestC" should match "remove -y TestE" with "Erase" and "3" packages # behave/steps/command_steps.py:153

  Scenario: History info                          # behave/history-2-dnf-2.feature:43
    When I save rpmdb                             # behave/steps/rpm_steps.py:16
    Then history info should match                # behave/steps/command_steps.py:195
      | Key          | Value            |
      | Command Line | install -y TestC |
      | Return-Code  | Success          |
      | Install      | TestC            |
    When I successfully run "dnf remove -y TestA" # behave/steps/command_steps.py:33
    Then history info should match                # behave/steps/command_steps.py:195
      | Key          | Value           |
      | Command Line | remove -y TestA |
      | Return-Code  | Success         |
      | Erase        | TestA, TestB    |

  Scenario: History info in range - transaction merging  # behave/history-2-dnf-2.feature:57
    When I save rpmdb                                    # behave/steps/rpm_steps.py:16
    And I successfully run "dnf install -y TestA"        # behave/steps/command_steps.py:33
    Then history info should match                       # behave/steps/command_steps.py:195
      | Key          | Value            |
      | Command Line | install -y TestA |
      | Return-Code  | Success          |
      | Install      | TestA, TestB     |
    When I save rpmdb                                    # behave/steps/rpm_steps.py:16
    And I enable repository "updates"                    # behave/steps/repo_steps.py:488
    And I successfully run "dnf update -y"               # behave/steps/command_steps.py:33
    Then history info should match                       # behave/steps/command_steps.py:195
      | Key          | Value        |
      | Command Line | update -y    |
      | Return-Code  | Success      |
      | Upgrade      | TestA, TestB |
      | Upgraded     | TestA, TestB |
    And history info "last-1..last" should match         # behave/steps/command_steps.py:195
      | Key         | Value   |
      | Return-Code | Success |
      | Install     | TestA   |
    And history info "last-2..last" should match         # behave/steps/command_steps.py:195
      | Key         | Value   |
      | Return-Code | Success |
      | Upgrade     | TestA   |
      | Upgraded    | TestA   |
    And history info "last-2..last-1" should match       # behave/steps/command_steps.py:195
      | Key         | Value   |
      | Return-Code | Success |
      | Reinstall   | TestA   |
      Assertion Failed: "TestA" not matched as "Reinstall"
      Captured stdout:
      shell.command: ['dnf-2', 'install', '-y', 'TestA']
      shell.command.stdout:
      Last metadata expiration check: 0:00:05 ago on Fri Oct  6 10:54:13 2017.
      Dependencies resolved.
      ================================================================================
       Package           Arch               Version            Repository        Size
      ================================================================================
      Installing:
       TestA             noarch             1-1                test             5.5 k
      Installing dependencies:
       TestB             noarch             1-1                test             5.4 k
      
      Transaction Summary
      ================================================================================
      Install  2 Packages
      
      Total size: 11 k
      Installed size: 0  
      Downloading Packages:
      Running transaction check
      Transaction check succeeded.
      Running transaction test
      Transaction test succeeded.
      Running transaction
        Preparing        :                                                        1/1 
        Installing       : TestB-1-1.noarch                                       1/2 
        Installing       : TestA-1-1.noarch                                       2/2 
        Verifying        : TestA-1-1.noarch                                       1/2 
        Verifying        : TestB-1-1.noarch                                       2/2 
      
      Installed:
        TestA.noarch 1-1                       TestB.noarch 1-1                      
      
      Complete!
      
      shell.command.stderr:
      
      shell.command: ['dnf-2', 'history', 'info']
      shell.command.stdout:
      Transaction ID : 10
      Begin time     : Fri Oct  6 10:54:18 2017
      Begin rpmdb    : 253:dc6011e64ea59d2d1b07e4dbe037b7c35d146418
      End time       : Fri Oct  6 10:54:19 2017 (1 seconds)
      End rpmdb      : 255:575e08a5768448b0f226c8f6d1288ef9cb4350a3
      User           : System <unset>
      Return-Code    : Success
      Command Line   : install -y TestA
      Transaction performed with:
          Installed     dnf-2.7.2-2gfc09f4b3.fc26.noarch @rpmsoftwaremanagement-dnf-nightly
          Installed     rpm-4.13.0.1-7.fc26.x86_64       @koji-override-0
      Packages Altered:
          Install TestA-1-1.noarch @test
          Install TestB-1-1.noarch @test
      
      shell.command.stderr:
      
      shell.command: ['dnf-2', 'update', '-y']
      shell.command.stdout:
      updates                                         1.1 MB/s | 1.1 kB     00:00    
      Last metadata expiration check: 0:00:00 ago on Fri Oct  6 10:54:19 2017.
      Dependencies resolved.
      ================================================================================
       Package           Arch               Version         Repository           Size
      ================================================================================
      Upgrading:
       TestA             noarch             2-1             updates             5.4 k
       TestB             noarch             2-1             updates             5.4 k
       TestC             noarch             2-1             updates             5.4 k
      
      Transaction Summary
      ================================================================================
      Upgrade  3 Packages
      
      Total size: 16 k
      Downloading Packages:
      Running transaction check
      Transaction check succeeded.
      Running transaction test
      Transaction test succeeded.
      Running transaction
        Preparing        :                                                        1/1 
        Upgrading        : TestC-2-1.noarch                                       1/6 
        Upgrading        : TestB-2-1.noarch                                       2/6 
        Upgrading        : TestA-2-1.noarch                                       3/6 
        Cleanup          : TestA-1-1.noarch                                       4/6 
        Cleanup          : TestB-1-1.noarch                                       5/6 
        Cleanup          : TestC-1-1.noarch                                       6/6 
        Verifying        : TestA-2-1.noarch                                       1/6 
        Verifying        : TestB-2-1.noarch                                       2/6 
        Verifying        : TestC-2-1.noarch                                       3/6 
        Verifying        : TestA-1-1.noarch                                       4/6 
        Verifying        : TestB-1-1.noarch                                       5/6 
        Verifying        : TestC-1-1.noarch                                       6/6 
      
      Upgraded:
        TestA.noarch 2-1          TestB.noarch 2-1          TestC.noarch 2-1         
      
      Complete!
      
      shell.command.stderr:
      
      shell.command: ['dnf-2', 'history', 'info']
      shell.command.stdout:
      Transaction ID : 11
      Begin time     : Fri Oct  6 10:54:19 2017
      Begin rpmdb    : 255:575e08a5768448b0f226c8f6d1288ef9cb4350a3
      End time       : Fri Oct  6 10:54:20 2017 (1 seconds)
      End rpmdb      : 255:ed0065aa62e7a94f1ce53a65359d8dc541d1d01c
      User           : System <unset>
      Return-Code    : Success
      Command Line   : update -y
      Transaction performed with:
          Installed     dnf-2.7.2-2gfc09f4b3.fc26.noarch @rpmsoftwaremanagement-dnf-nightly
          Installed     rpm-4.13.0.1-7.fc26.x86_64       @koji-override-0
      Packages Altered:
          Upgraded TestA-1-1.noarch @test
          Upgrade        2-1.noarch @updates
          Upgraded TestB-1-1.noarch @test
          Upgrade        2-1.noarch @updates
          Upgraded TestC-1-1.noarch @test
          Upgrade        2-1.noarch @updates
      
      shell.command.stderr:
      
      shell.command: ['dnf-2', 'history', 'info', 'last-1..last']
      shell.command.stdout:
      Transaction ID : 10..11
      Begin time     : Fri Oct  6 10:54:18 2017
      Begin rpmdb    : 253:dc6011e64ea59d2d1b07e4dbe037b7c35d146418
      End time       : Fri Oct  6 10:54:20 2017 (2 seconds)
      End rpmdb      : 255:ed0065aa62e7a94f1ce53a65359d8dc541d1d01c
      User           : System <unset>
      Return-Code    : Success
      Command Line   : install -y TestA
      Command Line   : update -y
      Transaction performed with:
          Installed     dnf-2.7.2-2gfc09f4b3.fc26.noarch @rpmsoftwaremanagement-dnf-nightly
          Installed     rpm-4.13.0.1-7.fc26.x86_64       @koji-override-0
      Packages Altered:
          Install  TestA-2-1.noarch @updates
          Install  TestB-2-1.noarch @updates
          Upgraded TestC-1-1.noarch @test
          Upgrade        2-1.noarch @updates
      
      shell.command.stderr:
      
      shell.command: ['dnf-2', 'history', 'info', 'last-2..last']
      shell.command.stdout:
      Transaction ID : 9..11
      Begin time     : Fri Oct  6 10:54:18 2017
      Begin rpmdb    : 253:dc6011e64ea59d2d1b07e4dbe037b7c35d146418
      End time       : Fri Oct  6 10:54:20 2017 (2 seconds)
      End rpmdb      : 255:ed0065aa62e7a94f1ce53a65359d8dc541d1d01c
      User           : System <unset>
      Return-Code    : Success
      Command Line   : install -y TestA
      Command Line   : remove -y TestA
      Command Line   : update -y
      Transaction performed with:
          Installed     dnf-2.7.2-2gfc09f4b3.fc26.noarch @rpmsoftwaremanagement-dnf-nightly
          Installed     rpm-4.13.0.1-7.fc26.x86_64       @koji-override-0
      Packages Altered:
          Upgraded TestA-1-1.noarch @test
          Upgrade        2-1.noarch @updates
          Upgraded TestB-1-1.noarch @test
          Upgrade        2-1.noarch @updates
          Upgraded TestC-1-1.noarch @test
          Upgrade        2-1.noarch @updates
      
      shell.command.stderr:
      
      shell.command: ['dnf-2', 'history', 'info', 'last-2..last-1']
      shell.command.stdout:
      Transaction ID : 9..10
      Begin time     : Fri Oct  6 10:54:18 2017
      Begin rpmdb    : 253:dc6011e64ea59d2d1b07e4dbe037b7c35d146418
      End time       : Fri Oct  6 10:54:19 2017 (1 seconds)
      End rpmdb      : 255:575e08a5768448b0f226c8f6d1288ef9cb4350a3
      User           : System <unset>
      Return-Code    : Success
      Command Line   : install -y TestA
      Command Line   : remove -y TestA
      Transaction performed with:
          Installed     dnf-2.7.2-2gfc09f4b3.fc26.noarch @rpmsoftwaremanagement-dnf-nightly
          Installed     rpm-4.13.0.1-7.fc26.x86_64       @koji-override-0
      Packages Altered:
      
      shell.command.stderr:


  Scenario: History info of package       # behave/history-2-dnf-2.feature:89
    And history info "TestA" should match # behave/steps/command_steps.py:195
      | Key         | Value                      |
      | Return-Code | Success                    |
      | Upgrade     | TestA, TestB               |
      | Upgraded    | TestA, TestB               |
      | Install     | TestA, TestB, TestC, TestE |
      | Erase       | TestA, TestB               |


Failing scenarios:
  behave/history-2-dnf-2.feature:57  History info in range - transaction merging

0 features passed, 1 failed, 0 skipped
6 scenarios passed, 1 failed, 0 skipped
31 steps passed, 1 failed, 0 skipped, 0 undefined
Took 0m8.315s
+ TEST_EXIT=1
+ '[' 0 == -r ']'
+ '[' 0 == -R -a 1 -ne 0 ']'
+ exit 1
```

**Comment by edynox** [comment]
This is kinda interesting. Transaction merging in DNF is not working correctly (works fine with SWDB). I tried it several times and found these problems:
`Assertion Failed: "TestA" not matched as "Reinstall"`

Caused by:
```
      Transaction ID : 9..10
      Begin time     : Mon Oct  9 08:41:30 2017
      Begin rpmdb    : 256:42c60d967d3cdfa5449883e8b04f406dc61c024d
      End time       : Mon Oct  9 08:41:31 2017 (1 seconds)
      End rpmdb      : 258:d32ee999405ceb17710925a2978babcc7a5c5b82
      User           : System <unset>
      Return-Code    : Success
      Command Line   : install -y TestA
      Command Line   : remove -y TestA
      Transaction performed with:
          Installed     dnf-2.6.4-0.32g3005c7ae.fc27.noarch  @rpmsoftwaremanagement-dnf-nightly
          Installed     rpm-4.13.90-0.git14002.6.fc27.x86_64 @anaconda
      Packages Altered:
```
Line
```
Reinstall TestA-1-1.noarch @test
```
is missing

**and**

`Assertion Failed: "TestA" not matched as "Upgraded"`

caused by

```
      Transaction ID : 9..11
      Begin time     : Mon Oct  9 08:39:44 2017
      Begin rpmdb    : 258:851d70c9b0f8505331baa72ce3e882d81791ac75
      End time       : Mon Oct  9 08:39:46 2017 (2 seconds)
      End rpmdb      : 258:9f4df76ce341884dfdcd3dc7718401079f3e7a95
      User           : System <unset>
      Return-Code    : Success
      Command Line   : remove -y TestA
      Command Line   : update -y
      Command Line   : install -y TestA
      Transaction performed with:
          Installed     dnf-2.6.4-0.32g3005c7ae.fc27.noarch  @rpmsoftwaremanagement-dnf-nightly
          Installed     rpm-4.13.90-0.git14002.6.fc27.x86_64 @anaconda
      Packages Altered:
          Reinstall TestA-1-1.noarch @test
          Upgrade   TestA-2-1.noarch @updates
          Reinstall TestB-1-1.noarch @test
          Upgrade   TestB-2-1.noarch @updates
          Upgraded  TestC-1-1.noarch @test
          Upgrade         2-1.noarch @updates
```
while it should be
```
          Upgraded TestA-1-1.noarch @test
          Upgrade        2-1.noarch @updates
          Upgraded TestB-1-1.noarch @test
          Upgrade        2-1.noarch @updates
          Upgraded TestC-1-1.noarch @test
          Upgrade        2-1.noarch @updates
```

**Comment by j-mracek** [comment]
@edynox Thanks, I will run tests several time just to ensure that it is fine.

**Comment by j-mracek** [comment]
Thanks, no problem discovered.

**Pull request merged**